WordPress/gutenberg
# [RNMobile] Fix undo/redo history when inserting a link configured to open in a new tab

## What?
Fixes a bug affecting undo/redo functionality on mobile. When a link is configured to open in a new tab and text is typed after this link, it is not possible to undo and then use the Redo button to redo the text that was added after the link.

**Original issue:**
* https://github.com/wordpress-mobile/gutenberg-mobile/issues/2091

**Related issues:**
* https://github.com/wordpress-mobile/gutenberg-mobile/issues/1160
* https://github.com/wordpress-mobile/gutenberg-mobile/issues/1161

**Gutenberg Mobile PR:**
* https://github.com/wordpress-mobile/gutenberg-mobile/pull/5747

## Why?
There are several related issues that describe the broken behavior with undo/redo when using the "Open in new tab" option on mobile. One symptom of this issue is that an extra state event occurs on mobile when a link is set to open in a new tab. This is demonstrated visually in https://github.com/wordpress-mobile/gutenberg-mobile/issues/2091#issuecomment-662212872 with further information. Configuring a link to open in a new tab shifts the `rel` attribute within the HTML, which causes the [edit reducer](https://github.com/WordPress/gutenberg/blob/trunk/packages/core-data/src/reducer.js#L523-L558) to disrupt the redo history.

Example HTML when a link is created without opening in a new tab:

```
<a href="https://wordpress.com" target="_blank" rel="noreferrer noopener">WordPress.com</a>
```

Example HTML when a link is created and configured to open in a new tab:

```
<a rel="noreferrer noopener" href="https://wordpress.com" target="_blank">WordPress.com</a>
```

## How?
This PR adds `rel` to the object of attributes passed to the link formatting, along with `url` and `target`:

```
attributes: {
	url: 'href',
	target: 'target',
	rel: 'rel',
},
```

When `rel` is passed with the attributes object to [registerFormatType](https://github.com/WordPress/gutenberg/blob/trunk/packages/format-library/src/index.js#L12), the `rel` attribute is no longer an unassigned attribute, and does not shift places when the HTML attributes are re-serialized. Therefore, the extra state event is not fired on mobile. On the web, the extra state event is not fired in either circumstance, and the re-serialization of the attributes has no effect on behavior. 

## Testing Instructions

1. In a post, add a paragraph block and type some text.
2. Create a link using the Link button in the Bottom Bar.
3. When configuring the link options, toggle the switch to **Open in new tab**.
4. Switch to HTML mode, and note the position of the `rel` attribute within the link's HTML. (It should be the final attribute.)
5. Switch back to Visual mode and continue typing a few more words of text.
6. Test the Undo and Redo buttons in the Bottom Bar.
7. Use the Undo button and step backward to at least the point before the link was created.
8. Use the Redo button to step forward to the point where the link was created, and also step forward to the text that was typed after the link.

**Demo builds for testing:**
- iOS: https://github.com/wordpress-mobile/WordPress-iOS/pull/20661
- Android: https://github.com/wordpress-mobile/WordPress-Android/pull/18405

## Screenshots

https://github.com/WordPress/gutenberg/assets/643285/bae80218-275c-4970-9e05-a37bb57088b6
